numpy/lib/npyio.py

```diff
@@ -1034,6 +1039,8 @@ def read_data(chunk_size):
             if len(X) > chunk_size:
                 yield X
                 X = []
+            if max_rows is not None and i >= max_rows-1:
```

Comment:
I think a clearer approach would be to change line 1022 to:
```
line_iter = itertools.chain([first_line], fh)
line_iter = itertools.islice(line_iter, max_rows)  # handles None correctly already
for i, line in line_iter:
    ...

Reply:
This is nicer. It seems `for i, line in enumerate(line_iter):` is needed

Reply:
Done as well